File: work/lib/a_stock_layer/strategies/portfolio/allocation.py
"""
Asset Allocation Models — 资产配置模型
=========================================
来源: Markowitz (MPT), Bridgewater (Risk Parity), Black-Litterman
"""
from __future__ import annotations

import numpy as np


def risk_parity_weights(cov_matrix: np.ndarray) -> np.ndarray:
    """风险平价权重 (Risk Parity) — 使每种资产的风险贡献度相等。"""
    n = cov_matrix.shape[0]
    inv_vol = 1.0 / np.sqrt(np.diag(cov_matrix))
    return inv_vol / inv_vol.sum()


def mean_variance_weights(expected_returns: np.ndarray,
                          cov_matrix: np.ndarray,
                          risk_aversion: float = 2.0) -> np.ndarray:
    """均值-方差最优权重 (Markowitz MPT)。"""
    n = len(expected_returns)
    inv_cov = np.linalg.inv(cov_matrix)
    ones = np.ones(n)
    # 简化版: 不约束做空
    w = np.dot(inv_cov, expected_returns) / risk_aversion
    return w / np.sum(w)


def kelly_fraction(win_rate: float, avg_win: float, avg_loss: float) -> float:
    """凯利公式最优投注比例。"""
    if avg_loss == 0:
        return 0
    b = avg_win / avg_loss
    q = 1 - win_rate
    f = (b * win_rate - q) / b
    return max(0, min(1, f))


def half_kelly(win_rate: float, avg_win: float, avg_loss: float) -> float:
    """半凯利 (更保守)。"""
    return kelly_fraction(win_rate, avg_win, avg_loss) * 0.5


def max_drawdown_limit(total_risk: float, max_dd: float = 0.15) -> float:
    """最大回撤限制下的仓位比例。"""
    if total_risk <= 0:
        return 0
    return min(1.0, max_dd / total_risk)


print("  ✔ portfolio/allocation")

File: packages/t-strategy/src/portfolio/allocation.ts
/** Asset allocation models */

export function meanVarianceWeights(
  expectedReturns: number[],
  covMatrix: number[][],
  riskAversion = 2.0,
): number[] {
  const n = expectedReturns.length
  if (n === 0 || covMatrix.length !== n) return []

  // Invert covariance via Gauss-Jordan (small n only)
  const inv = invertMatrix(covMatrix)
  if (!inv) return equalWeights(n)

  const raw = inv.map((row, i) =>
    row.reduce((sum, v, j) => sum + v * expectedReturns[j], 0) / riskAversion,
  )
  const total = raw.reduce((a, b) => a + b, 0)
  if (total === 0) return equalWeights(n)
  return raw.map(w => w / total)
}

function equalWeights(n: number) {
  return Array.from({ length: n }, () => 1 / n)
}

function invertMatrix(m: number[][]): number[][] | null {
  const n = m.length
  const aug = m.map((row, i) => {
    const ident = Array.from({ length: n }, (_, j) => (i === j ? 1 : 0))
    return [...row, ...ident]
  })

  for (let col = 0; col < n; col++) {
    let pivot = col
    for (let row = col + 1; row < n; row++) {
      if (Math.abs(aug[row][col]) > Math.abs(aug[pivot][col])) pivot = row
    }
    if (Math.abs(aug[pivot][col]) < 1e-12) return null
    if (pivot !== col) [aug[col], aug[pivot]] = [aug[pivot], aug[col]]

    const div = aug[col][col]
    for (let j = 0; j < 2 * n; j++) aug[col][j] /= div

    for (let row = 0; row < n; row++) {
      if (row === col) continue
      const factor = aug[row][col]
      for (let j = 0; j < 2 * n; j++) aug[row][j] -= factor * aug[col][j]
    }
  }

  return aug.map(row => row.slice(n))
}

export function riskParityWeights(covDiag: number[]) {
  const invVol = covDiag.map(v => (v > 0 ? 1 / Math.sqrt(v) : 0))
  const sum = invVol.reduce((a, b) => a + b, 0)
  return sum === 0 ? invVol : invVol.map(v => v / sum)
}

export function kellyFraction(winRate: number, avgWin: number, avgLoss: number) {
  if (avgLoss === 0) return 0
  const b = avgWin / avgLoss
  const q = 1 - winRate
  const f = (b * winRate - q) / b
  return Math.max(0, Math.min(1, f))
}

export function halfKelly(winRate: number, avgWin: number, avgLoss: number) {
  return kellyFraction(winRate, avgWin, avgLoss) * 0.5
}

export function maxDrawdownLimit(totalRisk: number, maxDd = 0.15) {
  if (totalRisk <= 0) return 0
  return Math.min(1, maxDd / totalRisk)
}
